### PR TITLE
Optimizing get indices to segments map calculations for IndicesSegmen…

### LINKED ISSUE #14720: [BUG] High number of unnecessary computations in getIndices() of IndicesSegmentResponse
### Describe the bug

For clusters with large number of shards and indices, the following computation is sub-optimal and leads to higher latencies for Rest APIs such as _cat/segments.

https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/action/admin/indices/segments/IndicesSegmentResponse.java#L96

### Related component

Cluster Manager

### To Reproduce

Create a cluster with high number of shards and indices (say 30k indices and 100k shards). 
The latency of ```_cat/segments``` API will increase manifold.

### Expected behavior

Refactoring using a map  ```Map<String, List<ShardSegments>> indicesShardSegments = new HashMap<>();``` will help save on extra computations and decrease latency of REST APIs.

### Additional Details

**Plugins**
Please list all plugins currently enabled.

**Screenshots**

**Host/Environment (please complete the following information):**

**Additional context**

### PR DESCRIPTION
**Pull request by gargharsh3134**
### Description
Removes extra computations to get the indices to IndexSegments map which will help reduce the latencies for _cat/segments and other REST APIs

### Related Issues
Resolves #14720 

### Check List
- [x] Functionality includes testing.
- [x] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md), if applicable.
- [x] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose), if applicable.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/src/main/java/org/opensearch/action/admin/indices/segments/IndicesSegmentResponse.java [hunk 1]
```diff
@@ -87,20 +85,19 @@ public Map<String, IndexSegments> getIndices() {
         }
         Map<String, IndexSegments> indicesSegments = new HashMap<>();
 
-        Set<String> indices = new HashSet<>();
+        Map<String, List<ShardSegments>> indicesShardSegments = new HashMap<>();
```
Comment (rajiv-kv) [inline server/src/main/java/org/opensearch/action/admin/indices/segments/IndicesSegmentResponse.java:88]: Wouldn't it now consume more memory - given that the map is retained for the lifetime of the method ?

File: server/src/main/java/org/opensearch/action/admin/indices/segments/IndicesSegmentResponse.java [hunk 2]
```diff
@@ -87,20 +85,19 @@ public Map<String, IndexSegments> getIndices() {
...
         for (ShardSegments shard : shards) {
```
Comment (rajiv-kv) [inline server/src/main/java/org/opensearch/action/admin/indices/segments/IndicesSegmentResponse.java:89]: If we sort the ShardSegments by `indexName` - can we avoid the double iteration ?

File: server/src/main/java/org/opensearch/action/admin/indices/segments/IndicesSegmentResponse.java [hunk 3]
```diff
@@ -85,22 +84,23 @@ public Map<String, IndexSegments> getIndices() {
...
+        Arrays.sort(shards, Comparator.comparing(shardSegment -> shardSegment.getShardRouting().getIndexName()));
+        Map<String, IndexSegments> indicesSegments = new HashMap<>();
 
-        for (String indexName : indices) {
-            List<ShardSegments> shards = new ArrayList<>();
-            for (ShardSegments shard : this.shards) {
-                if (shard.getShardRouting().getIndexName().equals(indexName)) {
-                    shards.add(shard);
-                }
+        int lastIdx = 0;
```
[the remark this answers is not in the document]
Comment (gargharsh3134) [reply server/src/main/java/org/opensearch/action/admin/indices/segments/IndicesSegmentResponse.java]: Updated.

File: server/src/main/java/org/opensearch/action/admin/indices/segments/IndicesSegmentResponse.java [hunk 4]
```diff
@@ -85,22 +84,23 @@ public Map<String, IndexSegments> getIndices() {
...
+        for (int i = 0; i < shards.length; i++) {
+            String lastIndexName = shards[lastIdx].getShardRouting().getIndexName();
```
[the remark this answers is not in the document]
Comment (gargharsh3134) [reply server/src/main/java/org/opensearch/action/admin/indices/segments/IndicesSegmentResponse.java]: Updated.

File: server/src/main/java/org/opensearch/action/admin/indices/segments/IndicesSegmentResponse.java [hunk 5]
```diff
@@ -85,22 +84,23 @@ public Map<String, IndexSegments> getIndices() {
...
+            String currentIndexName = shards[i].getShardRouting().getIndexName();
+            if (!currentIndexName.equals(lastIndexName)) {
+                indicesSegments.put(lastIndexName, new IndexSegments(lastIndexName, Arrays.copyOfRange(shards, lastIdx, i)));
+                lastIdx = i;
+            }
+            if (i == shards.length - 1) {
+                indicesSegments.put(currentIndexName, new IndexSegments(currentIndexName, Arrays.copyOfRange(shards, lastIdx, i + 1)));
```
[the remark this answers is not in the document]
Comment (gargharsh3134) [reply server/src/main/java/org/opensearch/action/admin/indices/segments/IndicesSegmentResponse.java]: Updated.

File: server/src/main/java/org/opensearch/action/admin/indices/segments/IndicesSegmentResponse.java [hunk 6]
```diff
@@ -86,21 +85,30 @@ public Map<String, IndexSegments> getIndices() {
...
+            if (!shards[i].getShardRouting().getIndexName().equals(startIndexName)) {
+                indicesSegments.put(startIndexName, new IndexSegments(startIndexName, Arrays.copyOfRange(shards, startIndexPos, i)));
+                startIndexPos = i;
+                startIndexName = shards[startIndexPos].getShardRouting().getIndexName();
             }
-            indicesSegments.put(indexName, new IndexSegments(indexName, shards.toArray(new ShardSegments[0])));
         }
+        // Add the last shardSegment from shards list which would have got missed in the loop above
+        indicesSegments.put(
+            shards[shards.length - 1].getShardRouting().getIndexName(),
```
Comment (rajiv-kv) [inline server/src/main/java/org/opensearch/action/admin/indices/segments/IndicesSegmentResponse.java:105]: nit:should this be `startIndexName ` ? the last batch that needs to be added ?

File: server/src/test/java/org/opensearch/action/admin/indices/segments/IndicesSegmentResponseTests.java [hunk 7]
```diff
@@ -68,4 +71,49 @@ public void testToXContentSerialiationWithSortedFields() throws Exception {
...
+                for (int segmentNum = 0; segmentNum < segmentsPerShard; segmentNum++) {
+                    segmentList.add(new Segment("foo" + indexName + shardId + segmentNum));
+                }
+                shardSegmentsList.add(new ShardSegments(shardRouting, segmentList));
+            }
+        }
+
+        // Prepare the IndicesSegmentResponse object and get the indicesSegments map
+        IndicesSegmentResponse response = new IndicesSegmentResponse(
+            shardSegmentsList.toArray(new ShardSegments[0]),
```
Comment (rajiv-kv) [inline server/src/test/java/org/opensearch/action/admin/indices/segments/IndicesSegmentResponseTests.java:101]: could we shuffle the array ?

File: server/src/test/java/org/opensearch/action/admin/indices/segments/IndicesSegmentResponseTests.java [hunk 8]
```diff
@@ -68,4 +71,49 @@ public void testToXContentSerialiationWithSortedFields() throws Exception {
...
+            totalIndices * shardsPerIndex,
+            totalIndices * shardsPerIndex,
+            0,
+            Collections.emptyList()
+        );
+        Map<String, IndexSegments> indicesSegments = response.getIndices();
+
+        assertEquals(totalIndices, indicesSegments.size());
```
Comment (rajiv-kv) [inline server/src/test/java/org/opensearch/action/admin/indices/segments/IndicesSegmentResponseTests.java:109]: Can we also assert for the values populated for  a segment of one shard ?

**Review by rajiv-kv — APPROVED** [review]
LGTM !

File: server/src/test/java/org/opensearch/action/admin/indices/segments/IndicesSegmentResponseTests.java (hunk 8 above)
Comment (gargharsh3134) [reply server/src/test/java/org/opensearch/action/admin/indices/segments/IndicesSegmentResponseTests.java:109]: Ack, will revise the UT.

File: server/src/main/java/org/opensearch/action/admin/indices/segments/IndicesSegmentResponse.java [hunk 9]
```diff
@@ -86,21 +85,24 @@ public Map<String, IndexSegments> getIndices() {
...
+            if (!shards[i].getShardRouting().getIndexName().equals(startIndexName)) {
+                indicesSegments.put(startIndexName, new IndexSegments(startIndexName, Arrays.copyOfRange(shards, startIndexPos, i)));
+                startIndexPos = i;
+                startIndexName = shards[startIndexPos].getShardRouting().getIndexName();
             }
-            indicesSegments.put(indexName, new IndexSegments(indexName, shards.toArray(new ShardSegments[0])));
         }
+        // Add the last shardSegment from shards list which would have got missed in the loop above
+        indicesSegments.put(startIndexName, new IndexSegments(startIndexName, Arrays.copyOfRange(shards, startIndexPos, shards.length)));
+
```
[the remark this answers is not in the document]
Comment (gargharsh3134) [reply server/src/main/java/org/opensearch/action/admin/indices/segments/IndicesSegmentResponse.java:105]: Yeah, we did consider it. The first implementation was using the Map based approach only ([conversation link](https://github.com/opensearch-project/OpenSearch/pull/14717#discussion_r1678820331)). But, then we decided to not use additional memory and keep the new implementation in line with the existing one where we are not creating the `Map<String, List<ShardSegments>>`, basically we traded for latency (not much difference) over memory.

The sort based approach is not creating any additional `Map<String, List<ShardSegments>> ` (which can be huge if the number of shards and indices are large), but incurring an additional cost of sorting (which should not be much, given that the shardCount will always be restricted in a sense) as compared to Map based approach which will require 2 passes plus an additional space.

Please let me know if you feel Map based appraoch is a better alternative.

[the remark this answers is not in the document]
Comment (gargharsh3134) [reply server/src/main/java/org/opensearch/action/admin/indices/segments/IndicesSegmentResponse.java:105]: > Do you have any numbers on the optimization achieved?

For cluster with ~500k shards and around 90TB data (~1,364,422 segments), the final latency comparison (with and without this change):


Stat | Without any optimisation (seconds) | With Optimisation (seconds)
-- | -- | --
Average | 120.22 | 11.94
P90 | 126.92 | 13.83
Max | 138.16 | 21.56
Min | 112.29 | 7.76

Though, don't have the comparison amongst the Map vs sort based approach.

[the remark this answers is not in the document]
Comment (backslasht) [reply server/src/main/java/org/opensearch/action/admin/indices/segments/IndicesSegmentResponse.java:105]: This helps, Thanks!

File: server/src/test/java/org/opensearch/action/admin/indices/segments/IndicesSegmentResponseTests.java (hunk 7 above)
Comment (gargharsh3134) [reply server/src/test/java/org/opensearch/action/admin/indices/segments/IndicesSegmentResponseTests.java:101]: Updated.

File: server/src/test/java/org/opensearch/action/admin/indices/segments/IndicesSegmentResponseTests.java (hunk 8 above)
Comment (gargharsh3134) [reply server/src/test/java/org/opensearch/action/admin/indices/segments/IndicesSegmentResponseTests.java:111]: Updated the UT. Thanks!

**Review by shwetathareja — APPROVED** [review]

**Comment by shwetathareja** [comment]
@gargharsh3134 observing this error
```
> Task :server:forbiddenApisTest
Forbidden method invocation: java.util.Collections#shuffle(java.util.List) [Use java.util.Collections#shuffle(java.util.List, java.util.Random) with a reproducible source of randomness]
  in org.opensearch.action.admin.indices.segments.IndicesSegmentResponseTests (IndicesSegmentResponseTests.java:99)

```

**Comment by gargharsh3134** [comment]
> @gargharsh3134 observing this error
> 
> ```
> > Task :server:forbiddenApisTest
> Forbidden method invocation: java.util.Collections#shuffle(java.util.List) [Use java.util.Collections#shuffle(java.util.List, java.util.Random) with a reproducible source of randomness]
>   in org.opensearch.action.admin.indices.segments.IndicesSegmentResponseTests (IndicesSegmentResponseTests.java:99)
> ```

Addressed in the latest commit. Thanks!

**Review by shwetathareja — APPROVED** [review]

**Pull request merged**